File: app/src/main/java/com/example/phonebook/dagger/modules/ViewModelsModule.java
package com.example.phonebook.dagger.modules;

import androidx.lifecycle.ViewModel;

import com.example.phonebook.dagger.scopes.ActivitiesScope;
import com.example.phonebook.dagger.scopes.ViewModelKeys;
import com.example.phonebook.viewModels.LogInViewModel;
import com.example.phonebook.viewModels.MainActivityViewModel;
import com.example.phonebook.viewModels.RegisterViewModel;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module
public abstract class ViewModelsModule {

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(LogInViewModel.class)
    public abstract ViewModel provideLogInViewModel(LogInViewModel logInViewModel);

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(MainActivityViewModel.class)
    public abstract ViewModel provideMainActivityViewModel(MainActivityViewModel mainActivityViewModel);

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(RegisterViewModel.class)
    public abstract ViewModel provideRegisterViewModel(RegisterViewModel registerViewModel);
}

File: app/src/main/java/com/example/phonebook/dagger/modules/ViewModelsModule.kt
package com.example.phonebook.dagger.modules

import androidx.lifecycle.ViewModel
import com.example.phonebook.dagger.scopes.ActivitiesScope
import com.example.phonebook.dagger.scopes.ViewModelKeys
import com.example.phonebook.ui.viewModels.*
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
abstract class ViewModelsModule {

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(AuthorisationActivityViewModel::class)
    internal abstract fun provideAuthorisationActivityViewModel(authorisationActivityViewModel: AuthorisationActivityViewModel): ViewModel

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(ContactCreationViewModel::class)
    internal abstract fun provideContactCreationViewModel(contactCreationViewModel: ContactCreationViewModel): ViewModel

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(ContactListViewModel::class)
    internal abstract fun provideContactListViewModel(contactListViewModel: ContactListViewModel): ViewModel

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(LogInViewModel::class)
    internal abstract fun provideLogInViewModel(logInViewModel: LogInViewModel): ViewModel

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(MainActivityViewModel::class)
    internal abstract fun provideMainActivityViewModel(mainActivityViewModel: MainActivityViewModel): ViewModel

    @Binds
    @IntoMap
    @ActivitiesScope
    @ViewModelKeys(RegisterViewModel::class)
    internal abstract fun provideRegisterViewModel(registerViewModel: RegisterViewModel): ViewModel

}